src/metadata/extra_tags.rs

```diff
@@ -0,0 +1,365 @@
+        self
+    }
+}
+
+/// The registry in which extra tags (parsers) are registered
+///
+/// Pass this to [`crate::metadata::TiffMetadataReader`] when reading.
+///
+/// ```
+/// # use async_tiff::reader::{AsyncFileReader, ObjectReader};
+/// # use async_tiff::metadata::TiffMetadataReader;
+/// use async_tiff::metadata::extra_tags::ExtraTagsRegistry;
+/// # use std::sync::Arc;
+/// # use std::env::current_dir;
+/// # use object_store::local::LocalFileSystem;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let registry = ExtraTagsRegistry::default();
+///
+///     let store = Arc::new(LocalFileSystem::new_with_prefix(current_dir().unwrap()).unwrap());
+///     # let path = "tests/sample-exif.tiff";
+///     let reader =
+///         Arc::new(ObjectReader::new(store.clone(), path.into())) as Arc<dyn AsyncFileReader>;
+///     let mut metadata_reader = TiffMetadataReader::try_open(&reader).await.unwrap();
+///     // get first ifd
+///     let ifd = &metadata_reader
+///         .read_all_ifds(&reader, registry)
+///         .await
+///         .unwrap()[0];
+///     // retrieve the registry
+///     println!("{:?}",ifd.extra_tags());
+/// }
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct ExtraTagsRegistry(HashMap<Tag, Arc<dyn ExtraTags>>);
+
+impl ExtraTagsRegistry {
+    /// Create a new, empty `ExtraTagsRegistry`
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+    /// checks if we have an entry for this tag
+    pub fn contains(&self, tag: &Tag) -> bool {
+        self.0.contains_key(tag)
+    }
+    /// Register an ExtraTags so their tags are parsed and stored in the ifd's `extra_tags``
+    pub fn register(&mut self, tags: Arc<dyn ExtraTags>) -> AsyncTiffResult<()> {
+        // check for duplicates
+        for tag in tags.tags() {
+            if self.0.contains_key(tag) {
+                return Err(AsyncTiffError::General(format!(
+                    "Tag {tag:?} already registered in {self:?}!"
+                )));
+            }
+        }
+        // add to self
+        for tag in tags.tags() {
+            self.0.insert(*tag, tags.clone());
+        }
+        Ok(())
+    }
+
+    /// deep clone so we have different registries per IFD
+    pub(crate) fn deep_clone(&self) -> Self {
+        let mut new_registry = ExtraTagsRegistry::new();
+
+        // we need to do some magic, because we can have multiple tags pointing to the same arc
+        let mut seen = HashSet::new();
+        for extra_tags in self.0.values() {
+            // only add if this is the first encountered reference to this arc
+            // (using thin pointer equality: https://stackoverflow.com/a/67114787/14681457 ; https://github.com/rust-lang/rust/issues/46139#issuecomment-346971153)
+            if seen.insert(Arc::as_ptr(extra_tags) as *const ()) {
+                if let Err(e) = new_registry.register(extra_tags.clone_arc()) {
+                    panic!("{e}");
+                }
+            }
```

Comment:
Since we only expect different arcs to come from `register`, which is `Arc::clone`, fat pointers should also compare equal, but there is a bit of a rabbit-hole with there being no guarantee on vtable layouts (IIRC)